jetty-server/src/main/java/org/eclipse/jetty/server/OptionalConnectionFactory.java

```diff
@@ -0,0 +1,225 @@
+                ((LifeCycle)connectionFactory).start();
+        }
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+
+        for (SensingConnectionFactory connectionFactory : _sensingConnectionFactories)
+        {
+            if (connectionFactory instanceof LifeCycle)
+                ((LifeCycle)connectionFactory).stop();
+        }
+    }
+
+    @Override
+    public Connection newConnection(Connector connector, EndPoint endPoint)
+    {
+        ConnectionFactory nextRecognizedConnectionFactory = getRecognizedConnectionFactory(connector, endPoint);
+        ConnectionFactory nextNotRecognizedConnectionFactory = getNotRecognizedConnectionFactory(connector, endPoint);
+        LOG.debug("Upgrade topology: from {} to {} on recognition, {} on non-recognition", getProtocol(), nextRecognizedConnectionFactory.getProtocol(), nextNotRecognizedConnectionFactory.getProtocol());
+        return configure(new OptionalConnection(endPoint, connector, _sensingConnectionFactories, nextRecognizedConnectionFactory, nextNotRecognizedConnectionFactory), connector, endPoint);
+    }
+
+    protected ConnectionFactory getRecognizedConnectionFactory(Connector connector, EndPoint endPoint)
+    {
+        String nextProtocol = _nextProtocol;
+        if (nextProtocol == null)
+            nextProtocol = SensingConnectionFactory.findNextProtocol(connector, getProtocol());
+        if (nextProtocol == null)
+            throw new IllegalStateException("No next protocol for " + endPoint);
+        return connector.getConnectionFactory(nextProtocol);
+    }
+
+    protected ConnectionFactory getNotRecognizedConnectionFactory(Connector connector, EndPoint endPoint)
+    {
+        return getRecognizedConnectionFactory(connector, endPoint);
+    }
+
+    static class OptionalConnection extends AbstractConnection implements Connection.UpgradeFrom
+    {
+        private final Connector _connector;
+        private final List<SensingConnectionFactory> _sensingConnectionFactories;
+        private final ConnectionFactory _nextRecognizedConnectionFactory;
+        private final ConnectionFactory _nextNotRecognizedConnectionFactory;
+        private final ByteBuffer _buffer;
+
+        protected OptionalConnection(EndPoint endp, Connector connector, List<SensingConnectionFactory> sensingConnectionFactories, ConnectionFactory nextRecognizedConnectionFactory, ConnectionFactory nextNotRecognizedConnectionFactory)
+        {
+            super(endp, connector.getExecutor());
+            _connector = connector;
+            _sensingConnectionFactories = sensingConnectionFactories;
+            _nextRecognizedConnectionFactory = nextRecognizedConnectionFactory;
+            _nextNotRecognizedConnectionFactory = nextNotRecognizedConnectionFactory;
+            _buffer = connector.getByteBufferPool().acquire(1024, true); // TODO make the buffer size & directility configurable
```

Comment:
Use `inputBufferSize` from `AbstractConnectionFactory`. Directness could be hardcoded (to `true`).

Reply:
Thanks for the pointer, I'll change the code accordingly.